File: src/data/tutorials/005-product-of-array-except-self.js
/**
 * Tutorial content for #005 — Product of Array Except Self
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an integer array, return a new array where each element at index i is the product of all elements in the original array except the one at index i. You must not use division, and the solution should run in O(n) time.`,
    example: `nums = [1, 2, 3, 4]\n→ result[0] = 2×3×4 = 24\n→ result[1] = 1×3×4 = 12\n→ result[2] = 1×2×4 = 8\n→ result[3] = 1×2×3 = 6\n✅ Answer: [24, 12, 8, 6]`,
    keyInsight: `For each position, the answer is (product of everything to the left) × (product of everything to the right). Compute these two products in two separate passes.`,
  },

  approaches: {
    'Brute Force': {
      intuition: `For each element, loop through the entire array and multiply all the other elements together. Skip the current index. Easy to understand but does a lot of repeated multiplication.`,
      steps: [
        `Create a result array of the same size.`,
        `Outer loop i from 0 to n-1.`,
        `Inner loop j from 0 to n-1: multiply all nums[j] where j ≠ i.`,
        `Store the product in result[i].`,
        `Return the result array.`,
      ],
      example: `nums = [1, 2, 3, 4]\n\ni=0: skip 0 → 2×3×4 = 24\ni=1: skip 1 → 1×3×4 = 12\ni=2: skip 2 → 1×2×4 = 8\ni=3: skip 3 → 1×2×3 = 6\n✅ Answer: [24, 12, 8, 6]`,
      keyInsight: `O(n²) time, O(1) extra space (ignoring output). Too slow for large arrays — for n=100,000 this is 10 billion multiplications.`,
    },

    'Division': {
      intuition: `Compute the total product of all elements, then divide by each element to get the "product of everything except self". This is simple and O(n), but the problem forbids division — and it also fails when any element is zero.`,
      steps: [
        `Compute totalProduct = product of all non-zero elements. Count zeros.`,
        `If there are 2+ zeros, every result is 0.`,
        `If there is exactly 1 zero, only the position of the zero gets a non-zero result.`,
        `If no zeros, result[i] = totalProduct / nums[i].`,
        `Return the result.`,
      ],
      example: `nums = [1, 2, 3, 4]\ntotalProduct = 1×2×3×4 = 24\n\nresult[0] = 24/1 = 24\nresult[1] = 24/2 = 12\nresult[2] = 24/3 = 8\nresult[3] = 24/4 = 6\n✅ Answer: [24, 12, 8, 6]\n\n(But nums=[1,0,3,4] would fail with plain division!)`,
      keyInsight: `O(n) time, O(1) extra space. Simple but requires careful zero-handling. Not allowed in the problem constraints — shown here to contrast with the preferred approach.`,
    },

    'Prefix + Suffix Arrays': {
      intuition: `Build two arrays: prefix[i] = product of all elements before index i, and suffix[i] = product of all elements after index i. The answer at each position is just prefix[i] × suffix[i]. Three clean passes over the array.`,
      steps: [
        `Build prefix[]: prefix[0]=1; for i≥1: prefix[i] = prefix[i-1] * nums[i-1].`,
        `Build suffix[]: suffix[n-1]=1; for i<n-1 going right-to-left: suffix[i] = suffix[i+1] * nums[i+1].`,
        `Build result[]: result[i] = prefix[i] * suffix[i].`,
        `Return result.`,
      ],
      example: `nums   = [1,  2,  3,  4]\nprefix = [1,  1,  2,  6]   (products before each index)\nsuffix = [24, 12, 4,  1]   (products after each index)\n\nresult[0] = 1  × 24 = 24\nresult[1] = 1  × 12 = 12\nresult[2] = 2  ×  4 = 8\nresult[3] = 6  ×  1 = 6\n✅ Answer: [24, 12, 8, 6]`,
      keyInsight: `O(n) time, O(n) space (two helper arrays). Very readable — the prefix/suffix pattern appears in many other problems so it's worth mastering.`,
    },

    'Optimized Space': {
      intuition: `Instead of two separate arrays, use the output array itself to store prefix products in the first pass, then multiply in the suffix products in a single reverse pass using just one running variable. This eliminates the need for the suffix array.`,
      steps: [
        `Initialize result[0]=1. Forward pass: result[i] = result[i-1] * nums[i-1] (prefix products in-place).`,
        `Initialize suffix=1. Reverse pass from i=n-1 down to 0:`,
        `  result[i] *= suffix  (multiply in the suffix product).`,
        `  suffix *= nums[i]    (extend the running suffix product).`,
        `Return result.`,
      ],
      example: `nums = [1, 2, 3, 4]\n\nForward (prefix into result):\nresult = [1, 1, 2, 6]\n\nReverse (multiply suffix in):\nsuffix=1\ni=3: result[3]=6*1=6,  suffix=1*4=4\ni=2: result[2]=2*4=8,  suffix=4*3=12\ni=1: result[1]=1*12=12, suffix=12*2=24\ni=0: result[0]=1*24=24, suffix=24*1=24\n✅ Answer: [24, 12, 8, 6]`,
      keyInsight: `O(n) time, O(1) extra space (only the output array). This is the interview-optimal answer — same logic as prefix+suffix but cleverly reuses the output array to cut space in half.`,
    },
  },
}

File: public/cpp/005-product-of-array-except-self.cpp
/*
 * LeetCode Problem #238: Product of Array Except Self
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/product-of-array-except-self/
 * 
 * Problem Statement:
 * Given an integer array nums, return an array answer such that answer[i] is
 * equal to the product of all the elements of nums except nums[i].
 * You must write an algorithm that runs in O(n) time and without using division.
 */

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    // ==================== APPROACH 1: Brute Force ====================
    /*
     * Algorithm:
     * - For each position i, multiply all elements except nums[i]
     * - Use nested loops
     * 
     * Time Complexity: O(n²)
     * Space Complexity: O(1) excluding output array
     * 
     * When to use: Never in interviews (too slow)
     */
    vector<int> productExceptSelf_BruteForce(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n);
        
        for (int i = 0; i < n; i++) {
            int product = 1;
            for (int j = 0; j < n; j++) {
                if (i != j) {
                    product *= nums[j];
                }
            }
            result[i] = product;
        }
        
        return result;
    }
    
    // ==================== APPROACH 2: Division (Not Allowed) ====================
    /*
     * Algorithm:
     * - Calculate total product of all elements
     * - For each position, divide total by nums[i]
     * - Handle zeros specially
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     * 
     * When to use: Problem explicitly forbids division
     * 
     * Note: This approach has issues with zeros and is not allowed
     */
    vector<int> productExceptSelf_Division(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n);
        
        // Count zeros and calculate product of non-zero elements
        int zeroCount = 0;
        long long product = 1;
        
        for (int num : nums) {
            if (num == 0) {
                zeroCount++;
            } else {
                product *= num;
            }
        }
        
        // Fill result based on zero count
        for (int i = 0; i < n; i++) {
            if (zeroCount > 1) {
                result[i] = 0;
            } else if (zeroCount == 1) {
                result[i] = (nums[i] == 0) ? product : 0;
            } else {
                result[i] = product / nums[i];
            }
        }
        
        return result;
    }
    
    // ==================== APPROACH 3: Prefix and Suffix Arrays ====================
    /*
     * Algorithm:
     * - Create prefix array: prefix[i] = product of all elements before i
     * - Create suffix array: suffix[i] = product of all elements after i
     * - result[i] = prefix[i] * suffix[i]
     * 
     * Time Complexity: O(n) - three passes
     * Space Complexity: O(n) - two extra arrays
     * 
     * When to use: Good intermediate solution
     * 
     * Key Insight:
     * - Product except self = (product of left) * (product of right)
     */
    vector<int> productExceptSelf_PrefixSuffix(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefix(n, 1);
        vector<int> suffix(n, 1);
        vector<int> result(n);
        
        // Build prefix array
        for (int i = 1; i < n; i++) {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }
        
        // Build suffix array
        for (int i = n - 2; i >= 0; i--) {
            suffix[i] = suffix[i + 1] * nums[i + 1];
        }
        
        // Calculate result
        for (int i = 0; i < n; i++) {
            result[i] = prefix[i] * suffix[i];
        }
        
        return result;
    }
    
    // ==================== APPROACH 4: Optimized Space (Optimal) ====================
    /*
     * Algorithm:
     * - Use result array to store prefix products
     * - Calculate suffix products on the fly and multiply with prefix
     * - Only use one extra variable for suffix product
     * 
     * Time Complexity: O(n) - two passes
     * Space Complexity: O(1) - excluding output array
     * 
     * When to use: This is the OPTIMAL solution
     * 
     * Key Insight:
     * - We can calculate suffix products while building result
     * - Use result array itself to store prefix products
     * - Use single variable to track running suffix product
     */
    vector<int> productExceptSelf_Optimal(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n, 1);
        
        // First pass: calculate prefix products and store in result
        int prefix = 1;
        for (int i = 0; i < n; i++) {
            result[i] = prefix;
            prefix *= nums[i];
        }
        
        // Second pass: calculate suffix products and multiply with result
        int suffix = 1;
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= suffix;
            suffix *= nums[i];
        }
        
        return result;
    }
};

// ==================== TEST CASES ====================
void printArray(const vector<int>& arr) {
    cout << "[";
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i];
        if (i < arr.size() - 1) cout << ",";
    }
    cout << "]" << endl;
}

void runTests() {
    Solution sol;
    
    // Test Case 1: Standard case
    vector<int> nums1 = {1, 2, 3, 4};
    cout << "Test 1: ";
    printArray(sol.productExceptSelf_Optimal(nums1));
    // Expected: [24,12,8,6]
    // Explanation: [2*3*4, 1*3*4, 1*2*4, 1*2*3]
    
    // Test Case 2: With zeros
    vector<int> nums2 = {-1, 1, 0, -3, 3};
    cout << "Test 2: ";
    printArray(sol.productExceptSelf_Optimal(nums2));
    // Expected: [0,0,9,0,0]
    
    // Test Case 3: Two elements
    vector<int> nums3 = {5, 2};
    cout << "Test 3: ";
    printArray(sol.productExceptSelf_Optimal(nums3));
    // Expected: [2,5]
    
    // Test Case 4: All ones
    vector<int> nums4 = {1, 1, 1, 1};
    cout << "Test 4: ";
    printArray(sol.productExceptSelf_Optimal(nums4));
    // Expected: [1,1,1,1]
    
    // Test Case 5: Negative numbers
    vector<int> nums5 = {-1, -2, -3, -4};
    cout << "Test 5: ";
    printArray(sol.productExceptSelf_Optimal(nums5));
    // Expected: [-24,-12,-8,-6]
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPARISON OF APPROACHES:
 * 
 * 1. Brute Force:
 *    Time: O(n²), Space: O(1)
 *    Too slow, not acceptable
 * 
 * 2. Division:
 *    Time: O(n), Space: O(1)
 *    Not allowed by problem constraints
 * 
 * 3. Prefix/Suffix Arrays:
 *    Time: O(n), Space: O(n)
 *    Good solution, easy to understand
 * 
 * 4. Optimized Space (RECOMMENDED):
 *    Time: O(n), Space: O(1)
 *    Optimal solution meeting all constraints
 * 
 * INTERVIEW TIPS:
 * - Clarify that division is not allowed
 * - Start with prefix/suffix approach
 * - Optimize to use result array for prefix
 * - Explain how suffix is calculated on the fly
 * - Mention that output array doesn't count as extra space
 * 
 * KEY INSIGHTS:
 * - Product except self = prefix product * suffix product
 * - We can reuse the output array to save space
 * - Calculate prefix forward, suffix backward
 * - No need to store suffix array, calculate on the fly
 * 
 * STEP-BY-STEP for [1,2,3,4]:
 * 
 * After prefix pass:
 * result = [1, 1, 2, 6]
 * (result[i] = product of all elements before i)
 * 
 * After suffix pass:
 * result = [24, 12, 8, 6]
 * (multiply each by product of all elements after i)
 * 
 * COMMON MISTAKES:
 * - Using division (not allowed)
 * - Not handling zeros correctly
 * - Counting output array as extra space
 * - Not optimizing from O(n) space to O(1)
 * 
 * FOLLOW-UP QUESTIONS:
 * - What if division was allowed? (Calculate total product, divide by each)
 * - How to handle multiple zeros? (Check zero count)
 * - Can we do it in one pass? (No, need two passes for prefix/suffix)
 * - What about overflow? (Use long long or modulo arithmetic)
 * 
 * RELATED PROBLEMS:
 * - Maximum Product Subarray (LeetCode #152)
 * - Product of the Last K Numbers (LeetCode #1352)
 */

// Made with Bob
